examples/CMakeLists.txt

```diff
@@ -0,0 +1,34 @@
+#############################
+## I. Set project details
+#############################
+cmake_minimum_required(VERSION 3.14)
+set(CMAKE_POLICY_DEFAULT_CMP0127 NEW)
+
+project("gate_benchmark"
+        VERSION 0.1.0
+        DESCRIPTION "Benchmark of parametric & non-parametric gates."
+        LANGUAGES CXX
+)
+
+#############################
+## II. Fetch project
+#############################
+
+Include(FetchContent)
+
+FetchContent_Declare(
+    Pennylane-lightning
+    GIT_REPOSITORY  https://github.com/PennyLaneAI/pennylane-lightning
+    GIT_TAG         master
+)
+FetchContent_MakeAvailable(Pennylane-lightning)
+
+#############################
+## III. Create project target
+#############################
+
+set(CMAKE_CXX_STANDARD 17) # At least C++17 is required
+# set(CMAKE_CXX_FLAGS "-O3") # Not including flags here as I pass them via commandline args in bash script (I would normally set it here)
```

Comment:
I think it is fine to set `O3` here. For real-world use cases, we don't want to pass compile time options through bash scripts, instead relying on the library we use to already be compiled with the optimal flags.

Reply:
Also, one thing we may/should do to ensure adherence with the rest of the library is make sure the clang-tidy option is run (and no errors given) at build-time. We strictly adhere to clang-tidy recommendations for code structure and formatting, treating all warnings as errors to avoid any future problems. The following is an excerpt from the Lightning CMake

```
option(ENABLE_WARNINGS "Enable warnings" ON)
option(ENABLE_CLANG_TIDY "Enable clang-tidy build checks" OFF)

if(ENABLE_CLANG_TIDY)
    if (NOT DEFINED CLANG_TIDY_BINARY)
        set(CLANG_TIDY_BINARY clang-tidy)
    endif()
    set(CMAKE_CXX_CLANG_TIDY ${CLANG_TIDY_BINARY};
                            -extra-arg=-std=c++17;
    )
endif()
```

which will build with support when enabled (`-DENABLE_CLANG_TIDY=ON`), and treat all warnings as errors.

Reply:
Dear @mlxd,
I have been working on adhering to the clang-tidy configuration that has been set up in PL lightning and added the suggested lines to the `CMakeLists.txt` file in the examples folder. I've experienced some difficulties while trying to CMake with `-DENABLE_CLANG_TIDY=ON` on macOS which I will outline:
  The macOS clang compiler does not come preinstalled with clang-tidy so I installed llvm@11 and llvm@13 via brew and  specify the clang-tidy binary by adding `DCLANG_TIDY_BINARY=/opt/homebrew/opt/llvm@11/bin/clang-tidy` when executing CMake. There is a problem with this setup which I cannot resolve: When executing `cmake -DENABLE_CLANG_TIDY=ON -DCLANG_TIDY_BINARY=/opt/homebrew/opt/llvm@11/bin/clang-tidy .. && make` in the master branch (same error in the examples folder) of PL lightning I get the error message 
  `{PATH_TO_PL_LIGHTNING}/pennylane-lightning/pennylane_lightning/src/simulator/StateVector.hpp:278:5: error: function 'getData' should be marked [[nodiscard]] [modernize-use-nodiscard,-warnings-as-errors]
      auto getData() const -> CFP_t * { return arr_; }
      ^
      [[nodiscard]] 
  8344 warnings generated.
  Suppressed 8351 warnings (8342 in non-user code, 9 NOLINT).
  Use -header-filter=.* to display errors from all non-system headers. Use -system-headers to display errors from system headers as well.
  1 warning treated as error
  make[2]: *** [pennylane_lightning/src/simulator/CMakeFiles/lightning_simulator.dir/StateVector.cpp.o] Error 1
  make[1]: *** [pennylane_lightning/src/simulator/CMakeFiles/lightning_simulator.dir/all] Error 2
  make: *** [all] Error 2`.
  
  I have tried explicitly including the `pennylane-lightning/src/.clang-tidy` config file using llvm@13 and the flag `--config-file` to no avail.

The only way I have been able to run clang-tidy on the `gate_benchmark.cpp` is to execute `cmake -DENABLE_CLANG_TIDY=on -DCLANG_TIDY_BINARY=/opt/homebrew/opt/llvm@13/bin/clang-tidy -DCMAKE_EXPORT_COMPILE_COMMANDS=ON ..` in the `/examples/build` directory as described here  https://stackoverflow.com/questions/39455090/clang-tidy-cant-find-my-header-files.
I then move the generated `compile_commands.json` file to the `examples` directory and execute clang-tidy on `gate_benchmark.cpp` which does not result in any errors. As this solution is sub-optimal I was wondering if you have any suggestions of what I could try next to make sure my code fulfils the clang-tidy configuration of PL lightning.

Reply:
Hey @isschoch thanks for your diligence on this. The clang-tidy warning (being reported as an error) is complaining about this line in lightning's `StateVector.hpp` file: https://github.com/PennyLaneAI/pennylane-lightning/blob/4d8df75da8b8c9f5576f464fb390a9b895988480/pennylane_lightning/src/simulator/StateVector.hpp#L278

Namely, the `[[nodiscard]]` qualifier (https://en.cppreference.com/w/cpp/language/attributes/nodiscard) means that the return type of the method must be captured. I think adding `[[nodiscard]]` similar to:
https://github.com/PennyLaneAI/pennylane-lightning/blob/4d8df75da8b8c9f5576f464fb390a9b895988480/pennylane_lightning/src/simulator/StateVector.hpp#L319
may fix the complaint. It may be possible to auto-fix it using the `-fix` and `-fix-error` arguments to clang-tidy on the Lightning source (see [here](https://stackoverflow.com/questions/53195335/clang-tidy-fix-does-not-apply-any-changes)).

Using CMake this way pulls the lightning repo into a `_deps` directory, which can be modified to test this out during compilation. If this fixes the problem, I think it would be good to add the missing statements as a separate Lightning PR, since this will likely happen again in future.

Reply:
Actually, on second thought, fixing in this PR should be fine, rather than opening a second one.

Reply:
Thank you for the quick and detailed reply. I got the same error at multiple locations so I also modified all of those. In addition clang-tidy requested that the namespaces be concatenated instead of nested in some parts of the code (i.e. `namespace Pennylane::Util{...}` instead of `namespace Pennylane{namespace Util{...}}`). Here's a quick overview of the code that caused the warnings:

https://github.com/isschoch/pennylane-lightning/blob/075671cf9a9c929fa062534ce21cce1fa99486c5/pennylane_lightning/src/algorithms/AdjointDiff.hpp#L139-L140

https://github.com/isschoch/pennylane-lightning/blob/075671cf9a9c929fa062534ce21cce1fa99486c5/pennylane_lightning/src/simulator/Gates.hpp#L15-L16

https://github.com/isschoch/pennylane-lightning/blob/075671cf9a9c929fa062534ce21cce1fa99486c5/pennylane_lightning/src/simulator/StateVector.hpp#L278

https://github.com/isschoch/pennylane-lightning/blob/075671cf9a9c929fa062534ce21cce1fa99486c5/pennylane_lightning/src/simulator/StateVectorManaged.hpp#L74

https://github.com/isschoch/pennylane-lightning/blob/075671cf9a9c929fa062534ce21cce1fa99486c5/pennylane_lightning/src/util/Util.hpp#L56-L57

I'll push the fixes to this PR if you give me the go ahead to keep it in the same PR (double checking since the fix isn't a one-liner anymore). With this clang-tidy doesn't give me any errors in the examples folder as long as I replace the `examples/build/_deps/pennylane-lightning-src/pennylane_lightning/src` with the fixed version.

Reply:
Nice work! I think pushing to this PR makes sense. If there are any issues, it will be caught by the automated CI test-suite. I'm happy to have a look at this PR once the changes are made too, and we can always adapt if needed.